metals/src/main/scala/scala/meta/internal/metals/OutlineFilesProvider.scala

```diff
@@ -0,0 +1,191 @@
+        } outlineFiles.putIfAbsent(
+          id,
+          new BuildTargetOutlineFilesProvider(
+            buildTargets,
+            buffers,
+            id,
+            wasSuccessful,
+          ),
+        )
+    }
+  }
+
+  def didChange(id: BuildTargetIdentifier, path: AbsolutePath): Unit = {
+    for {
+      provider <- outlineFiles.get(id)
+    } provider.didChange(path)
+  }
+
+  def getOutlineFiles(
+      buildTargetId: Option[BuildTargetIdentifier]
+  ): Optional[JOutlineFiles] = {
+    val res: Option[JOutlineFiles] =
+      for {
+        id <- buildTargetId
+        provider <- outlineFiles.get(id)
+        outlineFiles <- provider.outlineFiles()
+      } yield outlineFiles
+    res.asJava
+  }
+
+  def enrichWithOutlineFiles(
+      buildTargetId: Option[BuildTargetIdentifier]
+  )(vFile: CompilerVirtualFileParams): CompilerVirtualFileParams = {
+    val optOutlineFiles =
+      for {
+        id <- buildTargetId
+        provider <- outlineFiles.get(id)
+        outlineFiles <- provider.outlineFiles()
+      } yield outlineFiles
+
+    optOutlineFiles
+      .map(outlineFiles => vFile.copy(outlineFiles = Optional.of(outlineFiles)))
+      .getOrElse(vFile)
+  }
+
+  def enrichWithOutlineFiles(
+      path: AbsolutePath
+  )(vFile: CompilerVirtualFileParams): CompilerVirtualFileParams = {
+    val optOutlineFiles =
+      for {
+        bt <- buildTargets.inferBuildTarget(path)
+        provider <- outlineFiles.get(bt)
+        outlineFiles <- provider.outlineFiles()
+      } yield outlineFiles
+
+    optOutlineFiles
+      .map(outlineFiles => vFile.copy(outlineFiles = Optional.of(outlineFiles)))
+      .getOrElse(vFile)
+  }
+
+  def clear(): Unit = {
+    outlineFiles.clear()
+  }
+}
+
+class BuildTargetOutlineFilesProvider(
+    buildTargets: BuildTargets,
+    buffers: Buffers,
+    id: BuildTargetIdentifier,
+    wasCompilationSuccessful: Boolean,
+) {
+  private val changedDocuments =
+    ConcurrentHashSet.empty[AbsolutePath]
+
+  private val wasAllOutlined: AtomicBoolean =
+    new AtomicBoolean(false)
+
+  private val wasSuccessfullyCompiled: AtomicBoolean =
+    new AtomicBoolean(wasCompilationSuccessful)
+
+  def wasSuccessfullyCompiledByBuildServer: Boolean =
+    wasSuccessfullyCompiled.get()
+
+  def successfulCompilation(): Unit = {
+    wasSuccessfullyCompiled.set(true)
+    changedDocuments.clear()
+  }
+
+  def didChange(path: AbsolutePath): Boolean =
+    changedDocuments.add(path)
+
+  def outlineFiles(): Option[OutlineFiles] = {
+    if (!wasSuccessfullyCompiled.get() && !wasAllOutlined.getAndSet(true)) {
+      // initial outline compilation that is a substitute for build server compilation
+      val allFiles =
+        buildTargets
+          .buildTargetSources(id)
+          .flatMap(_.listRecursive.toList)
+          .flatMap(toVirtualFileParams(_))
+          .toList
+          .asJava
+      Some(
+        OutlineFiles(
+          allFiles,
```

Comment:
We may want to cap the number of files for initial compilation, so it won't take too long.

Reply:
We can do a high number like 300 or something